src/frontend/src/optimizer/plan_node/logical_filter.rs

```diff
@@ -185,11 +190,129 @@ impl ToBatch for LogicalFilter {
+                            if try_match_expand!(&now_expr.inputs()[0], ExprImpl::FunctionCall)?
+                                .get_expr_type()
+                                != Type::Now
+                            {
+                                return Err(ExprError::InvalidParam {
+                                    name: "now",
+                                    reason: String::from("now() must appear in the left side of the now delta expression directly"),
+                                }
+                                .into());
+                            }
+                        }
+                        _ => {
+                            return Err(ExprError::InvalidParam {
+                                name: "now",
+                                reason: String::from("now delta expression must be a trivial add/subtract expression"),
+                            }
+                            .into());
+                        }
+                    }
+                }
+            }
+
+            let mut now_conds = conjunctions
+                .drain_filter(|cond| cond.count_nows() > 0)
+                .map(|cond| {
+                    must_match!(cond, ExprImpl::FunctionCall(function_call) => {
+                        (convert_comparator_to_priority(function_call.get_expr_type()), function_call)
+                    })
+                })
+                .collect_vec();
+            now_conds.sort_by_key(|(comparator_priority, _)| *comparator_priority);
+            // We do simple logical filters first because it can reduce size of dynamic filter's
+            // cache.
+            let simple_logical = LogicalFilter::new(self.input(), Condition { conjunctions });
+            let mut cur_streaming = PlanRef::from(StreamFilter::new(
+                simple_logical.clone_with_input(new_input),
+            ));
+            // Rewrite each now condition. Replace `NowExpr` with `StreamNow` and replace
+            // `LogicalFilter` with `DynamicFilter`.
+            for (_, now_cond) in now_conds {
+                let left_index = must_match!(now_cond.inputs()[0], ExprImpl::InputRef(box ref input_ref) => input_ref.index());
+                let rht = must_match!(now_cond.inputs()[1], ExprImpl::FunctionCall(box ref function_call) => {
+                    match function_call.get_expr_type() {
+                        Type::Now => PlanRef::from(StreamNow::new(self.ctx())),
+                        Type::Add | Type::Subtract => {
+                            let mut now_delta_expr = function_call.clone();
+                            now_delta_expr.inputs_mut()[0] = ExprImpl::from(InputRef::new(0, DataType::Timestamp));
+                            // We cannot call `LogicalProject::to_stream()` here, because its input is already a stream.
+                            StreamProject::new(LogicalProject::new(StreamNow::new(self.ctx()).into(), vec![ExprImpl::from(now_delta_expr)])).into()
```

Comment:
The schema provided to `StreamNow` is wrong. `StreamNow`'s context should have a column called `now` in the schema.